### DISCUSSION TITLE
porting project with WROOM and PCM5102 to ESP32 A1S ES8388

### DISCUSSION THREAD

**Discussion by coatsiecoates** [pschatzmann/arduino-audio-driver#20] [Q&A] [answered]
I have an operational Arduino sketch (https://github.com/copych/AcidBox) that uses a WROOM module and PCM5102 DAC that I would like to use with the ESP32 A1S and omit the PCM5102. I have the audiokit board V2.2 with ES8388.

I've installed the Arduino Audio Driver and put in the setup:

  LOGLEVEL_AUDIODRIVER = AudioDriverInfo;
  CodecConfig cfg;
  cfg.input_device = ADC_INPUT_LINE1;
  cfg.output_device = DAC_OUTPUT_ALL;
  cfg.i2s.bits = BIT_LENGTH_16BITS;
  cfg.i2s.rate = RATE_44K;
  AudioKitEs8388V2.begin(cfg);

I commented out the sketch's existing initialization so as not to conflict with the Arduino Audio Driver.

I'm using the headphone out jack to a powered speaker, which worked when I tried the sine generator example using audiotools.

The sketch compiles and uploads, but I'm getting no output.

The readme file says "After setting up the codec, you can use the I2S functionality provided by your microcontroller" so I am only assuming to leave the rest of the sketch intact, though obviously, I'm missing something. Anyone have any ideas?

If anyone decides to try it, it's important that the config.h has enabled:

#define JUKEBOX
#define JUKEBOX_PLAY_ON_START

**Comment by pschatzmann** [comment, marked as answer]
Did you get any error messages ?
I guess you forgot to define the i2s pins. 

I was using the same [scenario with Mozzi](https://www.pschatzmann.ch/home/2023/02/25/using-the-audiokit-library-with-mozzi/)...

**Reply by coatsiecoates** [reply]
I did not get any error messages, it compiled and uploaded. I went to AudioConfigESP32.h and added the correct information. Still no sound. I'm wondering if I'm placing the stuff in the correct places. 

I put:

LOGLEVEL_AUDIODRIVER = AudioDriverInfo;
CodecConfig cfg;
cfg.input_device = ADC_INPUT_LINE1;
cfg.output_device = DAC_OUTPUT_ALL;
cfg.i2s.bits = BIT_LENGTH_16BITS;
cfg.i2s.rate = RATE_44K;
AudioKitEs8388V2.begin(cfg);

directly at the beginning of the setup. I also trying the sketch's i2sInit(); before and after all of that.

I put #include "AudioBoard.h" as the first include in the sketch, but noticed that  #include "driver/i2s.h" is also called from AudioBoardESP32.h, but in <> and not "" so I moved #include AudioBoard after the sketch's include. Still nothing. When I commented out the already existing call to driver/i2s.h, I got my first errors.

I also noticed the silkscreen on my audiokit board is ESP32 Audio Kit V2.2 A247, only 3 digits unlike the 2957 or 3478 boards in the setup table. Perhaps this could be an issue.

**Reply by coatsiecoates** [reply]
An unrelated thing of note: if I click the link you provided in your reply, the results are different. A march 10 2023 post on my phone, and a february 25 2023 post if I click on desktop.

**Comment by pschatzmann** [comment]
I think the first step should be to find out what board you have ! This will give you the right I2S pins to use 

Try [this audiotools sketch](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-generator-audiokit/streams-generator-audiokit.ino) to check if it is working. 

I don't know what you refer to with AudioBoardESP32.h

**Reply by pschatzmann** [reply]
And don't forget to define [the master clock pin](https://github.com/pschatzmann/arduino-audiokit/wiki/GPIO-Overview-by-Selected-Board): w/o this it will not work!
Can you share your  pin setup ?

**Reply by coatsiecoates** [reply]
yep. The audiotools sketch you linked to works fine.

It's getting very very close to working. I have some recognizable sound now.

My apologies for  writing AudioBoardESP32.h, I meant AudioConfigESP32.h

// Set output mode
//#define ESP32_AUDIO_OUT_MODE INTERNAL_DAC
#define ESP32_AUDIO_OUT_MODE I2S_DAC_AND_I2S_ADC

// For external I2S output, only: I2S_PINS
#define ESP32_I2S_BCK_PIN 27
#define ESP32_I2S_WS_PIN 25
#define ESP32_I2S_DATA_PIN 26
#define ESP32_I2S_DATA_PIN_IN 35

Not sure where to define the master clock pin. The only reference I can find to it is to the old Arduino ADF/AudioKit HAL

I renamed all the existing defines to match the AudioConfigESP32.h names and I'm getting something that resembles the synth sounds I'm used to with the 5102 but it's choppy. Tried changing the sample rate but no change in performance.

the following is the setup called by the sketch with the changes made, but still no master clock, as I don't know where to put it, but it's getting very close! Should I delete any of this that might be conflicting with AudioBoard.h? SAMPLE_RATE is 44100, I tried 44000 as well

void i2sInit() {
  i2s_config_t i2s_config = {
    .mode = (i2s_mode_t)(I2S_MODE_MASTER | I2S_MODE_TX | I2S_MODE_RX),     //////  I added the RX
    .sample_rate = SAMPLE_RATE,
    .bits_per_sample = I2S_BITS_PER_SAMPLE_16BIT,
    .channel_format = I2S_CHANNEL_FMT_RIGHT_LEFT,
    .communication_format = (i2s_comm_format_t)(I2S_COMM_FORMAT_STAND_I2S ),
    .intr_alloc_flags = ESP_INTR_FLAG_LEVEL2,
    .dma_buf_count = DMA_NUM_BUF,
    .dma_buf_len = DMA_BUF_LEN,
    .use_apll = true,
  };

  i2s_pin_config_t i2s_pin_config = {
    .bck_io_num = ESP32_I2S_BCK_PIN,
    .ws_io_num =  ESP32_I2S_WS_PIN,
    .data_out_num = ESP32_I2S_DATA_PIN,
    .data_in_num = ESP32_I2S_DATA_PIN_IN
  };

  i2s_driver_install(i2s_num, &i2s_config, 0, NULL);

  i2s_set_pin(i2s_num, &i2s_pin_config);
  i2s_zero_dma_buffer(i2s_num);

  // DEBF("I2S is started: BCK %d, WCK %d, DAT %d\r\n", I2S_BCLK_PIN, I2S_WCLK_PIN, I2S_DOUT_PIN);
}

**Reply by coatsiecoates** [reply]
I tried this but it's not changing anything:

In config.h:

#define ESP32_I2S_BCK_PIN    27      // I2S BIT CLOCK pin (BCL BCK CLK)
#define ESP32_I2S_WS_PIN    25      // I2S WORD CLOCK pin (WCK WCL LCK)
#define ESP32_I2S_DATA_PIN    26      // to I2S DATA IN pin (DIN D DAT)
#define ESP32_I2S_DATA_PIN_IN    35
#define ESP32_I2S_MCLK_PIN    0

In setup routine:

 i2s_pin_config_t i2s_pin_config = {
    .mck_io_num = ESP32_I2S_MCLK_PIN,
    .bck_io_num = ESP32_I2S_BCK_PIN,
    .ws_io_num =  ESP32_I2S_WS_PIN,
    .data_out_num = ESP32_I2S_DATA_PIN,
    .data_in_num = ESP32_I2S_DATA_PIN_IN
  };

**Reply by pschatzmann** [reply]
Did you confirm with the sine test sketch that these are the correct pins ?
Not sure if the sequency is important, but mck_io_num should usually be last...

I think if you set the log level to debug the logger shows the pins that are effectively used.

AudioKitEs8388V2.begin(cfg); pins are 
```C++
    // add i2s pins: mclk, bck, ws,data_out, data_in ,(port)
    addI2S(PinFunction::CODEC, 0, 5, 25, 26, 35);
```
Which does not match with your pins!

So I am asking again which version did you manage to have working with my test sketch (AudioKitEs8388V1, AudioKitEs8388V2 or AudioKitAC101) ?

**Reply by coatsiecoates** [reply]
I just now figured it out. It was the V2 that worked, I tried the mck_io_num last and got errors, it would only compile if I entered it first. It was actually working many hours ago as it turns out. The choppiness in the audio had to do with something in the way the sequencer timing worked . As soon as I commented out /*read_button(&buttons[i]);*/ in the sequencer, the audio worked perfectly. Strange that it worked fine on a WROOM with no PSRAM but caused a problem with the A1S. Thank you so much for your time. I never would have figured it out without your help. Your work is so vast, it would have taken a month of study just to get an idea.
